iceoryx_posh/source/popo/ports/client_port_user.cpp

```diff
@@ -40,45 +39,78 @@ ClientPortUser::MemberType_t* ClientPortUser::getMembers() noexcept
 }
 
 cxx::expected<RequestHeader*, AllocationError>
-ClientPortUser::allocateRequest(const uint32_t /*userPayloadSize*/) noexcept
+ClientPortUser::allocateRequest(const uint32_t userPayloadSize, const uint32_t userPayloadAlignment) noexcept
 {
-    return cxx::error<AllocationError>(AllocationError::RUNNING_OUT_OF_CHUNKS);
+    auto allocateResult = m_chunkSender.tryAllocate(
+        getUniqueID(), userPayloadSize, userPayloadAlignment, sizeof(RequestHeader), alignof(RequestHeader));
+
+    if (allocateResult.has_error())
+    {
+        return cxx::error<AllocationError>(allocateResult.get_error());
+    }
+
+    auto requestHeader = new (allocateResult.value()->userHeader())
+        RequestHeader(getMembers()->m_uniqueId, RpcBaseHeader::UNKNOWN_CLIENT_QUEUE_INDEX);
+
+    return cxx::success<RequestHeader*>(requestHeader);
 }
 
-void ClientPortUser::freeRequest(RequestHeader* const /*requestHeader*/) noexcept
+void ClientPortUser::freeRequest(RequestHeader* const requestHeader) noexcept
 {
-    /// @todo
+    m_chunkSender.release(requestHeader->getChunkHeader());
 }
 
-void ClientPortUser::sendRequest(RequestHeader* const /*requestHeader*/) noexcept
+void ClientPortUser::sendRequest(RequestHeader* const requestHeader) noexcept
 {
-    /// @todo
+    const auto connectRequested = getMembers()->m_connectRequested.load(std::memory_order_relaxed);
+
+    if (connectRequested)
+    {
+        m_chunkSender.send(requestHeader->getChunkHeader());
+    }
+    else
+    {
+        LogWarn() << "Try to send request without being connected!";
+    }
 }
 
 void ClientPortUser::connect() noexcept
 {
-    /// @todo
+    if (!getMembers()->m_connectRequested.load(std::memory_order_relaxed))
```

Comment:
Can't we just store `true` here? After `load` and before `store` the value of `m_connectRequested` could have been changed.

Reply:
It can not change since it is only read on the RouDi side and on the Client side it is not multi-threaded. I more or less copied the code from `PublisherPortUser`/`SubscriberPortUser` but noticed that the `SubscriberPortUser` cleans up the receiver queue when the state changes. I guess it should be the same for the `ClientPortUser`. What do you think?